Rotate EC2 subnets on capacity constraints

rbrw opened the pull request.

rbrw committed:
aws_sdk: move wait_for_status() to launch_all_nodes()

rbrw committed:
aws_sdk: add create_instance()

Move the per-host core of launch_all_nodes() to create_instance() to
prepare for the addition of launch exception
handlers (InsufficientCapacity, etc.).

rbrw committed:
aws_sdk: Kill EC2 instances on exception

Move EC2 cleanup code to kill_instances() and call that from cleanup(),
and from a new exception handler in launch_all_nodes(), to be sure we
don't leave instances running when something goes wrong.

Adjust wait_for_status() to handle a specific list of hosts rather than
just @hosts.

rbrw commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -220,125 +221,176 @@ def kill_zombie_volumes
+      {instance: region.instances.create(config),
+       amitype: amitype,
+       amisize: amisize,
+       image_type: image_type}
```
@kbarber pointed out that this probably isn't idiomatic formatting, so I'll change the brace placement, indentation, etc. in the next version.  He also mentioned that the "foo:" form may not be supported by all the relevant versions of Ruby (so I'll see about that as well).

kbarber replied:
Yeah, its not support in Ruby 1.8.7 I believe, and I think we still support that. At least, its still being tested in travis-ci: https://github.com/puppetlabs/beaker/blob/master/.travis.yml

kbarber commented:
@rbrw do we have a job somewhere that shows this working with PuppetDB? Would we need to see that before we should raise this up, at least for our own assurances that it works in our pipeline with all of our hosts perhaps? I'm just wary of unforeseen circumstances that aren't picked up until we merge, that might be disruptive.

kbarber commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -69,26 +66,30 @@ def provision
+    # It goes without saying, but a #cleanup does nothing without a #provision
+    # method call first.
+    #
+    # @return [unspecified]
```
[void] is good enough, quote from yarddoc: http://www.rubydoc.info/gems/yard/file/docs/Tags.md#Literals

rbrw replied:
@kbarber ahh, I see.  Thanks.

kbarber commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -220,125 +221,176 @@ def kill_zombie_volumes
+            else
+              info = create_instance(host, ami_spec, subnet)
+            end
+            instances.push({instance: info[:instance], host: host})
```
Same incompat here, keyword/colon format ...

kbarber commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -220,125 +221,176 @@ def kill_zombie_volumes
+        :subnet => subnet_id,
+      }
+
+      {instance: region.instances.create(config),
```
If the response format is different for this method, then the yarddoc should be changed to reflect that, right now its still a [void].

kbarber commented:
@rbrw you can setup travis on your own repo, and actually have this repo run the Ruby rspec tests for you automatically if you want, saves having to find out when you raise the final PR.

kbarber commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -220,125 +221,176 @@ def kill_zombie_volumes
+      #
+      # Q: is this the correct way to look for something that's
+      # supposed to be in CONFIG?
+      subnets = @hosts[0]['subnets']
```
So here's the modelling conundrum ... a 'subnet_id' per host was allowed in the past so that if a user wanted to provide _different_ subnet_id's per node, they could. This obviously kind of overrides that. Now we don't use this today mind you, but the theory is that one might want to raise 2 nodes in very different data centres to test attenuated connections.

For a more global setting, that isn't per node, I think we'll need to look at @options['subnets'] also. This mechanism allows for providing a set of subnet id's for the whole yaml file, instead of per node.

rbrw replied:
@kbarber OK, thanks - I'd gotten the impression that the CONFIG settings were shoved down into each host by default, and hadn't noticed @options[].  If that lets us ask about CONFIG settings directly, that'd be better.

kbarber commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -220,125 +221,176 @@ def kill_zombie_volumes
+      end
+      ami_spec = YAML.load_file(@options[:ec2_yaml])["AMI"]
+      instances = [] # Each element is {instance: i, host: h}
+      subnets.each do |subnet|
```
I wonder if we should do a random shuffle first, to avoid always hitting a/b/c in order ... just if "a" has capacity issues, every test will be attempting to raise and then failing etc. its an interesting thing to ponder at least, not that important today.

kbarber replied:
On the 'different config per node' comment I had before, I wonder if we should invert this, and instead of having an all or nothing approach, just fall through on a per node basis. I see very little worries about having nodes launched in different subnet_id's (besides the attenuation I guess) and if we try the subnets in order, most will fall into the same subnet anyway.

This actually avoids the need to do an entire teardown for all nodes if just 1 node provision failed for example. Its an inversion of the way we're doing it now, and it gets us back to per node configuration again also.

rbrw replied:
@kbarber good idea with respect to shuffling.  Regarding subnet straddles, I suspect you and @senior might want to chat, since I implemented the current behavior after asking him, Wyatt, etc. about it.

kbarber replied:
So to be clear, this is a modelling bug as it stands I feel. And I can certainly see how a global setting is better for us then a per node, totally in agreement here. But, we have an existing API to maintain also ... so we need to make a decision about the direction (global versus not) but we still have to support those users using the old mechanism, and perhaps deprecate that functionality and retire in time. This code is published and public etc, and we can't just break it on a minor release - so we need to be thoughtful about that.

rbrw replied:
Right, I assumed we needed to discuss it.  In any case, now that we've decided to drop the all-or-nothing behavior, I think I have a solution that should be more or less completely backward compatible.

rbrw commented:
@kbarber I'm still waiting on a full test run, but I think this version should address most if not all of your suggestions.

kbarber commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -220,125 +221,171 @@ def kill_zombie_volumes
+        raise RuntimeError, 'Config specifies both subnet_id and subnet_ids'
+      end
+      any_subnet_hosts, other_hosts = @hosts.partition do |host|
+        global_subnets && ! (global_subnet_id || host['subnet_id'])
```
would a single node subnet_id not override a global? In line 231 we have this precedence ordering already, might be surprising.

rbrw replied:
@kbarber the intent here is that subnet_ids will only take effect for a host if subnet_ids has been specified in the config, no global subnet_id has been specified, and no host-specific subnet has been specified.  I'm not sure I follow what you mean regarding line 231.

rbrw replied:
@kbarber Sounded like line 355 handled things the way you expected.

kbarber commented on lib/beaker/hypervisor/aws_sdk.rb:
```diff
@@ -220,125 +221,171 @@ def kill_zombie_volumes
+          instance = create_instance(host, ami_spec, subnet_id)
+          instances.push({:instance => instance, :host => host})
+        end
+        wait_for_status(:running,
```
Is the new api for wait_for_status reasonable? I mean, we prepare this array of hashes, only to split it back into two arrays, which we later combine back later within wait_for_status using zip. Seems like a bit of back and forth?

rbrw replied:
@kbarber I don't disagree, but wasn't sure at the time (with respect to Ruby) it'd be palatable to have it accept the somewhat internal sequence of {:instance, :host} values instead.  Would that be preferable?

rbrw replied:
@kbarber I changed it to do that in the new version.

rbrw commented:
@kbarber OK, I've pushed a new version that may address your concerns.

rbrw commented:
The tests are being run via https://github.com/puppetlabs/puppetdb/pull/1359

rbrw committed:
aws_sdk: given CONFIG subnet_ids, try each in turn

This should help alleviate the EC2 InsufficientInstanceCapacity.

The new policy will only be applied to a host if there isn't a CONFIG
subnet_id and there isn't a host-specific subnet_id.  When it does
apply, try each subnet in subnet_ids once for each relevant host.
